temporian/implementation/numpy/operators/set_index.py

```diff
@@ -76,28 +75,26 @@ def _append_impl(event: NumpyEvent, append_feat_names: List[str]) -> NumpyEvent:
     }
     # initialize destination event & sampling data
     dst_event_data = {}
-    for src_idx_lvl, src_idx_lvl_data in event.iterindex():
-        dst_idx_suffs = [
-            tuple(x)
-            for x in zip(*[src_idx_lvl_data.features[i] for i in dst_idx_pos])
-        ]
-        for target_dst_idx_suff in set(dst_idx_suffs):
-            # find all occurrences of destination suff in source event
-            dst_idx_suff_pos = [
-                idx
-                for idx, dst_idx_suff in enumerate(dst_idx_suffs)
-                if dst_idx_suff == target_dst_idx_suff
-            ]
+    for src_idx_key, src_idx_data in event.iterindex():
+        # constructing the dict of unique tuples and their positions
+        dst_idx_suffs = defaultdict(list)
+        for i, tpl in enumerate(
+            zip(*[src_idx_data.features[j] for j in dst_idx_pos])
+        ):
+            tpl = tuple(tpl)
+            dst_idx_suffs[tpl].append(i)
+
+        for target_dst_idx_suff, dst_idx_suff_pos in dst_idx_suffs.items():
             # create destination index
-            dst_idx_lvl = src_idx_lvl + tuple(target_dst_idx_suff)
+            dst_idx_lvl = src_idx_key + tuple(target_dst_idx_suff)
 
             dst_event_data[dst_idx_lvl] = IndexData(
                 [
-                    event[src_idx_lvl].features[feat_pos][dst_idx_suff_pos]
-                    for feat_name, feat_pos in dst_feat_pos.items()
+                    src_idx_data.features[feat_pos][dst_idx_suff_pos]
+                    for feat_pos in dst_feat_pos.values()
```

Comment:
Would this code be more readable using only feature idx (i.e. positions) everywhere. For example, "dst_feat_pos" is a dictionary. For instance, "dst_feat_pos:list[int]".

In addition, while this is now defined in python, I think it would be better if we did not rely on the iteration order of dictionaries. This would improve readability.